BUG: Fix error message for unknown methodtype

yeesian opened the pull request:
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

yeesian committed:
BUG: Fix error message for unknown methodtype

phofl reviewed (commented):
Nice! Can you add a test?

yeesian commented:
Hi sorry for the slow response and thanks for the ping! I wasn't sure where to be adding the test and what it should look like. Was investigating for a bit but then work caught up so this slipped by, sorry again for stalling here, maybe some pointers can help me

MarcoGorelli commented:
sure

for a start, could you come a up with a small reproducible example which reproduces the issue which you're fixing?

yeesian committed:
Add test for unknown methodtype

yeesian committed:
Stay within column limits

MarcoGorelli commented on pandas/tests/test_errors.py:
```diff
@@ -110,3 +110,8 @@ def test_AbstractMethodError_classmethod():
     with pytest.raises(AbstractMethodError, match=xpr):
         Foo().method()
+
+    types = {"method", "classmethod", "staticmethod", "property"}
+    xpr = f"methodtype must be one of {types}, got unknownmethod instead."
```
let's avoid logic in tests if we can - can you write the raw string please?

ideally, it would probably be good to have an error message like
```
'methodtype' must be one of {'method', 'classmethod', 'staticmethod', 'property'}, got 'unknownmethod' instead
```

yeesian replied:
https://github.com/pandas-dev/pandas/pull/51484/commits/b697aba60edcfade44c09e904e85c7aa5663711b

MarcoGorelli reviewed (changes requested):
nice, getting there

could you add a whatsnew note to the 'other enhancements' section please? there should be a section on error reporting

yeesian committed:
Try interpolating the types into the error message

yeesian committed:
Add entry in announcements

MarcoGorelli commented on pandas/tests/test_errors.py:
```diff
@@ -110,3 +110,8 @@ def test_AbstractMethodError_classmethod():
         Foo().method()
+
+    xpr = "methodtype must be one of {'method', 'classmethod', "
+          "'staticmethod', 'property'}, got unknownmethod instead."
```
final nitpick, the rest looks good (pending green)
```
got 'unknownmethod' instead
```
(same for where the error is raised)

yeesian replied:
https://github.com/pandas-dev/pandas/pull/51484/commits/58a54bae6292cff2b61ea3996db59c24fc62f779

yeesian committed:
Add quotation marks

MarcoGorelli commented:
thanks for updating!

could you run the test locally please to check it passes? there are also some pre-commit failures, please check the contributing guide for how to run them

yeesian commented:
Sorry for the slow updates here, just to give a brief update:

* I tried following https://pandas.pydata.org/docs/dev/development/contributing_codebase.html#running-the-test-suite but ran into the following error:

```python
$ pytest pandas/tests/test_errors.py
ImportError while loading conftest '/Users/yeesian/Projects/pandas/pandas/conftest.py'.
pandas/__init__.py:22: in <module>
    from pandas.compat import is_numpy_dev as _is_numpy_dev  # pyright: ignore # noqa:F401
pandas/compat/__init__.py:25: in <module>
    from pandas.compat.numpy import (
pandas/compat/numpy/__init__.py:4: in <module>
    from pandas.util.version import Version
pandas/util/__init__.py:2: in <module>
    from pandas.util._decorators import (  # noqa:F401
pandas/util/_decorators.py:14: in <module>
    from pandas._libs.properties import cache_readonly
pandas/_libs/__init__.py:13: in <module>
    from pandas._libs.interval import Interval
E   ModuleNotFoundError: No module named 'pandas._libs.interval'
```

I tried looking up the error, got some vague advice about uninstalling and reinstalling pandas and gave up.

I think the precommit failures might be coming from the indentation in https://github.com/pandas-dev/pandas/pull/51484/commits/b697aba60edcfade44c09e904e85c7aa5663711b: I'll be happy to revert the change in https://github.com/pandas-dev/pandas/pull/51484#discussion_r1148406692 and interpolate the types into the error message just to keep things moving (I personally don't think it's that much logic and it runs the risk of being flakey about the way that sets should be displayed as strings), otherwise I'll rather bail on this PR.

yeesian committed:
Attempt at addressing precommit failures

yeesian commented:
It's failing on ARM in CircleCI (see below) but passing with the rest:

```
=================================== FAILURES ===================================
_____________________ test_AbstractMethodError_classmethod _____________________
[gw0] linux -- Python 3.8.15 /home/circleci/miniconda3/envs/pandas-dev/bin/python3.8

    def test_AbstractMethodError_classmethod():
        xpr = "This classmethod must be defined in the concrete class Foo"
        with pytest.raises(AbstractMethodError, match=xpr):
            Foo.classmethod()
    
        xpr = "This property must be defined in the concrete class Foo"
        with pytest.raises(AbstractMethodError, match=xpr):
            Foo().property
    
        xpr = "This method must be defined in the concrete class Foo"
        with pytest.raises(AbstractMethodError, match=xpr):
            Foo().method()
    
        types = {"method", "classmethod", "staticmethod", "property"}
        xpr = f"methodtype must be one of {types}, got 'unknownmethod' instead."
        with pytest.raises(AbstractMethodError, match=xpr):
>           Foo().unknownmethod()
E           AttributeError: 'Foo' object has no attribute 'unknownmethod'
```

yeesian committed:
Add corresponding method

yeesian commented:
It's unclear to me why [the initialization of `AbstractMethodError` requires a `methodtype: str`](https://github.com/pandas-dev/pandas/blob/d69eaae79af408b1a941ec42f6ac52a545bc39c0/pandas/errors/__init__.py#L190) to begin with, given that [it's not used anywhere else in the codebase](https://github.com/search?q=repo%3Apandas-dev%2Fpandas+AbstractMethodError&type=code) so I think this PR is addressing hypothetical scenarios that users don't actually hit in practice. A better PR might be to simply remove the argument from the constructor altogether.

yeesian committed:
The interpolation of sets into the error message was out-of-order

Given that we just wanted to test that the codepath is exercised, I followed https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

yeesian committed:
Fix the regex

* `\{`: Matches the opening brace
* `[^}]*`: Matches any number of characters that are not the closing brace
* `\}`: Matches the closing brace

yeesian committed:
Inline the set into the message

Gave up trying to get the regex right, and followed the feedback in https://github.com/pandas-dev/pandas/pull/51484#discussion_r1148406692 to write the raw string.

yeesian committed:
Fix precommit error

MarcoGorelli commented:
> A better PR might be to simply remove the argument from the constructor altogether.

yeah, sure, if it's actually unused that sounds good to me

MarcoGorelli commented:
closing as stale, but thanks for your PR - feel free to open a new PR if you'd like to continue
